lib/ProductOpener/Food.pm

```diff
@@ -1260,21 +1265,45 @@ normalize_serving_size(2.5kg)              returns 2500
 =cut
 
 sub normalize_serving_size($) {
-
 	my $serving = shift;
 
-	my $q = 0;
-	my $u;
-
-	if ($serving =~ /((\d+)(\.|,)?(\d+)?)( )?($units)\b/i) {
-		$q = lc($1);
-		$u = $6;
+	# Regex captures any <number>( )?<unit-identifier> group, but leaves allowances for a preceding
+	# token to allow for patterns like "One bag (32g)", "1 small bottle (180ml)" etc
+	if ($serving =~ /^(.*[ \(])?(?<quantity>(\d+)(\.|,)?(\d+)?)( )?(?<unit>\w+)\b/i) {
+		my $q = $+{quantity};
+        my $u = normalize_unit($+{unit});
 		$q = convert_string_to_number($q);
-		$q = unit_to_g($q,$u);
+
+		return unit_to_g($q, $u);
 	}
 
 	#$log->trace("serving size normalized", { serving => $serving, q => $q, u => $u }) if $log->is_trace();
-	return $q;
+	return 0;
+}
+
+# @todo we should have equivalences for more units if we are supporting this
+my %unit_equivalence_map = (
```

Comment:
Maybe instead of a hash, we could do a list, as we will iterate over all values anyway.
That way we could have multiple regexps that correspond to a single unit.
That could be useful if we automatically add regexps from translations.

Something like:

```
my @unit_equivalences = (
('g', qr/gram(s)?/),
('g',qr/gramme(s)?/), # French
);
```

Reply:
Agreed. I think this is more scalable and optimal.